### PR TITLE
Show photos taken in the visible area of the map

### PR DESCRIPTION
**Pull request by RaymondHuang210129**
First of all, Thanks for your great project! This app allows me to migrate my photos from Google to my self-hosted NAS without compromising the user experience.

I like the location feature in Google Photos because I can locate my travel photos by dragging the map to specific places.
Thus I would like to contribute some codes to make this feature available in this app.

I added another page called 'Locations' which shows the map on the top, and the timeline shows the photos taken within the visible range in the map.

I guess the only way to change schema automatically is to have the app updated and run the migration script, so I manually added two columns (latitude, longitude) to the table `oc_memories` and runs the below query to fill each row:

```sql
ALTER TABLE oc_memories ADD latitude double DEFAULT NULL;
ALTER TABLE oc_memories ADD longitude double DEFAULT NULL;
UPDATE IGNORE oc_memories SET longitude = IF(exif REGEXP '(?<=GPSLatitude\":).*?(?=,)' AND exif REGEXP '(?<=GPSLongitude\":).*?(?=,)', REGEXP_SUBSTR(exif, '(?<=GPSLongitude\":).*?(?=,)') + 0.0, NULL);
UPDATE IGNORE oc_memories SET latitude = IF(exif REGEXP '(?<=GPSLatitude\":).*?(?=,)' AND exif REGEXP '(?<=GPSLongitude\":).*?(?=,)', REGEXP_SUBSTR(exif, '(?<=GPSLatitude\":).*?(?=,)') + 0.0, NULL);
```
The IGNORE statement results from the photos with malformed GPSLatitude/Longitude values taken by some devices (my old HTC phone treats it as string type instead of rational64u when the value is close to 0).

In my test bed with 13k photos, The webpage does not slow down after applying the geometric filter.

I did not implement the marker clustering feature on the map yet, since client-side clustering requires all marker information sent to the frontend and it could cause bad performance (as discussed in #121 ). I didn't find any server-side clustering library for PHP yet, so I think I should implement this feature in the next step.

BTW, My VSCode linter doesn't seem to match your coding style. Could you provide such configurations in README? Thanks a lot.

### CODE REVIEW

File: lib/Controller/LocationsController.php [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+class LocationsController extends ApiBase
+{
+    use FoldersTrait;
+
+    /**
+     * @NoAdminRequired
+     *
+     * @PublicPage
+     */
+    public function daysWithBounds(string $minLat, string $maxLat, string $minLng, string $maxLng): JSONResponse
```
Comment (pulsejet) [inline lib/Controller/LocationsController.php:39]: This code is duplicating the Days code. It should programmed as a transform instead.

File: lib/Db/TimelineQueryDays.php [hunk 2]
```diff
@@ -124,6 +124,63 @@ public function getDays(
...
+     * @param bool         $archive         Whether to get the days only from the archive folder
+     * @param array        $queryTransforms An array of query transforms to apply to the query
+     * @param string       $minLat          The minimum latitude
+     * @param string       $maxLat          The maximum latitude
+     * @param string       $minLng          The minimum longitude
+     * @param string       $maxLng          The maximum longitude
+     *
+     * @return array The days response
+     */
+    public function getDaysWithBounds(
```
Comment (pulsejet) [inline lib/Db/TimelineQueryDays.php:141]: Again, this needs to be a transform, not duplicate.

**Review by pulsejet — CHANGES_REQUESTED** [review]
Hi @RaymondHuang210129, sorry for the delay in reviewing this.

> I added another page called 'Locations' which shows the map on the top, and the timeline shows the photos taken within the visible range in the map.

Overall this seems like a good start for implementing the maps feature. I'd like to note here that this will have to be hidden behind a "experimental" flag until the rest of the features are completed: otherwise it'll be confusing UX.

> I guess the only way to change schema automatically is to have the app updated and run the migration script, so I manually added two columns (latitude, longitude) to the table `oc_memories` and runs the below query to fill each row:
> 
> ```sql
> ALTER TABLE oc_memories ADD latitude double DEFAULT NULL;
> ALTER TABLE oc_memories ADD longitude double DEFAULT NULL;
> UPDATE IGNORE oc_memories SET longitude = IF(exif REGEXP '(?<=GPSLatitude\":).*?(?=,)' AND exif REGEXP '(?<=GPSLongitude\":).*?(?=,)', REGEXP_SUBSTR(exif, '(?<=GPSLongitude\":).*?(?=,)') + 0.0, NULL);
> UPDATE IGNORE oc_memories SET latitude = IF(exif REGEXP '(?<=GPSLatitude\":).*?(?=,)' AND exif REGEXP '(?<=GPSLongitude\":).*?(?=,)', REGEXP_SUBSTR(exif, '(?<=GPSLatitude\":).*?(?=,)') + 0.0, NULL);
> ```
> 
> The IGNORE statement results from the photos with malformed GPSLatitude/Longitude values taken by some devices (my old HTC phone treats it as string type instead of rational64u when the value is close to 0).

I would not worry about this. Asking users to re-run indexing after a major upgrade is fine.

> In my test bed with 13k photos, The webpage does not slow down after applying the geometric filter.

I don't expect it to. One thing to check would be the query response times (compared to the plain timeline queries). If everything is okay they should be very similar.

> I did not implement the marker clustering feature on the map yet, since client-side clustering requires all marker information sent to the frontend and it could cause bad performance (as discussed in #121 ). I didn't find any server-side clustering library for PHP yet, so I think I should implement this feature in the next step.

Sounds good.

> BTW, My VSCode linter doesn't seem to match your coding style. Could you provide such configurations in README? Thanks a lot.

Just use Prettier for Vue and TS. For PHP, you need to run `make php-lint`

File: lib/Controller/LocationsController.php (hunk 1 above)
Comment (RaymondHuang210129) [reply lib/Controller/LocationsController.php:39]: Could you explain more about programmed as a transform? Do you mean polymorphism, or is this a language feature specific to PHP?

**Comment by RaymondHuang210129** [comment]
Refactored the code as a transform filter.

**Comment by pulsejet** [comment]
@RaymondHuang210129 I should've dropped a note earlier: I'm working on something similar right now for map clustering. Thankfully I think there's not a lot of conflict yet so I'm going to try merging in your code locally and continue from there. But I'll suggest putting this on hold for the time being to avoid duplicate efforts.

**Comment by RaymondHuang210129** [comment]
@pulsejet Got it, thanks! I will wait for your merging process. Sorry for knowing this message late.

**Pull request closed without merging**